File: src/main/java/com/kingmang/lazurite/modules/Async/Async.java
package com.kingmang.lazurite.modules.Async;

import com.kingmang.lazurite.base.Arguments;
import com.kingmang.lazurite.modules.Module;
import com.kingmang.lazurite.runtime.*;
import com.kingmang.lazurite.runtime.ClassValue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class Async implements Module {
    public void init() {
        Map<String, Value> async = new HashMap<>();
        async.put("supply", new FunctionValue((args -> {
            Arguments.check(1, args.length);
            CompletableFuture<NumberValue> new_async;
            new_async = CompletableFuture.supplyAsync(() -> {
                (((FunctionValue) args[0]).getValue()).execute();
                return NumberValue.ONE;
            });
            return NumberValue.MINUS_ONE;
        })));
        newClass("async", new ArrayList<>(), async);
    }
    private static void newClass(String name, List<String> structArgs, Map<String, Value> targets) {
        ClassValue result = new ClassValue(name, structArgs);
        for (Map.Entry<String, Value> entry : targets.entrySet()) {
            Value expr = entry.getValue();
            result.setField(entry.getKey(), expr);
        }

        Variables.set(name, result);
    }
}

File: src/main/java/com/kingmang/lazurite/libraries/lzr/utils/async/async.kt
package com.kingmang.lazurite.libraries.lzr.utils.async

import com.kingmang.lazurite.core.Arguments
import com.kingmang.lazurite.core.Function
import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.libraries.Library
import com.kingmang.lazurite.runtime.Variables
import com.kingmang.lazurite.runtime.values.LzrFunction
import com.kingmang.lazurite.runtime.values.LzrMap
import com.kingmang.lazurite.runtime.values.LzrNumber
import java.util.concurrent.CompletableFuture

@Suppress("unused", "ClassName")
class async : Library {
    override fun init() {
        Variables.define("async", async())
    }

    override fun provides(): MutableMap<String, Pair<Int, Function>> {
        val map: MutableMap<String, Pair<Int, Function>> = HashMap()
        map["async"] = Pair(Types.MAP, Function { async() })
        return map
    }

    private fun async(): LzrMap {
        val async = LzrMap(1)
        async["supply"] = Function { args ->
            Arguments.check(1, args.size)
            CompletableFuture.supplyAsync {
                (args[0] as LzrFunction).value.execute()
                LzrNumber.ONE
            }
            LzrNumber.MINUS_ONE
        }
        return async
    }
}
